File: java/src/main/java/com/qingsong/leetcode/Problem1105.java
package com.qingsong.leetcode;

import java.lang.Math;


class Problem1105 {
    public int minHeightShelves(int[][] books, int shelfWidth) {
        int[] states = new int[books.length + 1];
        for (int i = 0; i < books.length; ++i) {
            states[i + 1] = states[i] + books[i][1];
        }
        
        for (int i = 0; i < books.length; ++i) {
            int floorMaxHeight = 0;
            int leftWidth = shelfWidth;
            for (int j = i; j > -1; --j) {
                leftWidth -= books[j][0];
                if (leftWidth < 0) {
                    break;
                }
                floorMaxHeight = Math.max(floorMaxHeight, books[j][1]);
                states[i + 1] = Math.min(states[i + 1], states[j] + floorMaxHeight);
            }
        }
        return states[books.length];
    }
}

File: python/leetcode/problem_1105.py
# 题目链接：https://leetcode.cn/problems/filling-bookcase-shelves/
from typing import List


def minHeightShelves(books: List[List[int]], shelfWidth: int) -> int:
    bookCount = len(books)
    states = [0]
    for book in books:
        states.append(states[-1] + book[1])

    for i in range(bookCount):
        floorMaxHeight, leftWidth = 0, shelfWidth
        for j in range(i, -1, -1):
            leftWidth -= books[j][0]
            if leftWidth < 0:
                break
            floorMaxHeight = max(floorMaxHeight, books[j][1])
            states[i + 1] = min(states[i + 1], states[j] + floorMaxHeight)
    return states[bookCount]
